libraries/AP_IBus_Telem/AP_IBus_Telem.h

```diff
@@ -0,0 +1,130 @@
+#define IBUS_MEAS_TYPE_ACC_Y          0x0d // Acc Y
+#define IBUS_MEAS_TYPE_ACC_Z          0x0e // Acc Z
+#define IBUS_MEAS_TYPE_ROLL           0x0f // Roll
+#define IBUS_MEAS_TYPE_PITCH          0x10 // Pitch
+#define IBUS_MEAS_TYPE_YAW            0x11 // Yaw
+#define IBUS_MEAS_TYPE_VERTICAL_SPEED 0x12 // Vertical speed
+#define IBUS_MEAS_TYPE_GROUND_SPEED   0x13 // Speed m/s
+#define IBUS_MEAS_TYPE_GPS_DIST       0x14 // Distance from home
+#define IBUS_MEAS_TYPE_ARMED          0x15 // Armed / unarmed (1 = armed, 0 = unarmed)
+#define IBUS_MEAS_TYPE_FLIGHT_MODE    0x16 // Flight mode (see Mode.h for flightmodes in copter)
+
+#define IBUS_MEAS_TYPE_PRES           0x41 // Pressure
+#define IBUS_MEAS_TYPE_SPE            0x7e // Speed km/h
+
+class AP_IBus_Telem
+{
+public:
+    // constructor
+    AP_IBus_Telem() {}
+
+    /* Do not allow copies */
+    AP_IBus_Telem(const AP_IBus_Telem &other) = delete;
+    AP_IBus_Telem &operator=(const AP_IBus_Telem &) = delete;
+
+    void init();
+    void loop();
+
+private:
+    AP_HAL::UARTDriver *_port;
+    bool _initialized;
+
+    void setSensorMeasurement(uint8_t adr, int32_t value); // sets a sensor measurement for a certain address
+    void setSensorMeasurements(); // set new values for all sensors configured
+    void populate_checksum(uint8_t *packet, uint16_t size); //populates the last two bytes with checksum
+    void handleTelemetryMessage(); // handles the telemetry message
+
+    enum class State {
+        GET_LENGTH,
+        GET_DATA,
+        GET_CHKSUML,
+        GET_CHKSUMH,
+        DISCARD
+    };
+
+    static const uint8_t PROTOCOL_MAX_LENGTH = 0x08;           // maximum length for sensor frames - 8 bytes
+    static const uint8_t PROTOCOL_OVERHEAD = 3;            // packet is <len><cmd><data....><chkl><chkh>, overhead=cmd+chk bytes
+    static const uint8_t PROTOCOL_COMMAND_DISCOVER = 0x80; // Command discover sensor (lowest 4 bits are sensor)
+    static const uint8_t PROTOCOL_COMMAND_TYPE = 0x90;     // Command discover sensor (lowest 4 bits are sensor)
+    static const uint8_t PROTOCOL_COMMAND_VALUE = 0xA0;    // Command send sensor data (lowest 4 bits are sensor)
+    static const uint8_t PROTOCOL_TIMEGAP = 3; // Packets are received very ~7ms so use ~half that for the gap
+    static const uint16_t UPDATE_MEASUREMENTS_TIMING = 500; // every 500ms
+    static const uint8_t PROTOCOL_FOUR_LENGTH = 0x04;      // indicate that the message has 4 bytes
+    static const uint8_t PROTOCOL_SIX_LENGTH = 0x06;      // indicate that the message has 6 bytes
+
+    State _state;
+    uint32_t _last; // milis() of prior message
+    uint32_t _last_send_measurements_time; // in millis time when the last time measurements were updated
+    uint8_t _buffer[PROTOCOL_MAX_LENGTH];
+    uint8_t _offs;     // pointer in buffer
+    uint8_t _len;     // incoming message length
+    uint16_t _chksum; // checksum calculation
+    uint8_t _lchksum;           // checksum lower byte received
```

Comment:
Do without this one and just use `_chksum`

Reply:
@peterbarker would it make sense to remove them all and put them as variables in the loop (update) function?